### PR TITLE
[dygraph hybrid pp for interleave] The interleave scheduler for pipeline parallel

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR types
Others

### PR changes
Others

### Describe
Support interleave scheduler.
Main changes:
1. Interleave scheduler for both pipeline training and evaluation.
2. Under eager dygraph mode, use async comm methods (isend/irecv) instead of sync methods (send/recv).
3. Add two new methods `_get_p2p_next_rank()`, `_get_p2p_prev_rank()` in class `HybridCommunicateGroup` to get the global rank of prev pp stage and next pp stage.
4. In p2p communication, use the global rank to represent the prev pp stage and next pp stage instead of only use `0` or `1` to represent.

**Note that: the interleave scheduler only supports eager dygraph mode.**

All works for supporting interleave scheduler for pipeline:
- [x] Support virtual pipeline layer spit (https://github.com/PaddlePaddle/Paddle/pull/45402)
- [x] Support virtual pipeline layer forward function (compatible with recompute) (https://github.com/PaddlePaddle/Paddle/pull/45444)
- [x] Implement interleave scheduler in pipeline_parallel (https://github.com/PaddlePaddle/Paddle/pull/45497)
- [ ] NCCLGroup start and NCCLGroup end. Implement save/load function for pipeline_parallel with virtual stage
- [ ] All necessary UT relating with recompute, AMP and etc.
- [ ] Some other functions but not sure yet

### CODE REVIEW

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 1]
```diff
@@ -53,12 +104,15 @@ def __init__(self, layers, hcg, strategy):
         self.stage_id = self._hcg.get_stage_id()
         self.pp_group = self._hcg.get_pipe_parallel_group()
 
+        global _REAL_PP_WORLD_SIZE
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: Why we must need a global variable?

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 2]
```diff
@@ -22,11 +22,47 @@
...
 __all__ = []
 
+_VIRTUAL_PP_WORLD_SIZE = None
+_VIRTUAL_PP_RANK = None
+
+_REAL_PP_WORLD_SIZE = None
+_REAL_PP_RANK = None
+
+
+def is_pipeline_first_stage(ignore_virtual=False):
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: I think we may split 1F1B pipeline and interleave pipe into two class.

**Review by gongweibao — CHANGES_REQUESTED** [review]
I think we may split 1F1B pipeline and interleave pipe into two class.

File: python/paddle/distributed/fleet/meta_parallel/pp_utils/p2p_communication.py [hunk 3]
```diff
@@ -323,105 +322,124 @@ def _p2p_helper(tensor_send_next, tensor_send_prev, recv_prev, recv_next):
...
-                         group=_hcg.recv_next_group,
-                         use_calc_stream=True)
-
-            allgather_partial(tensor_recv_next,
-                              nranks=mp_degree,
-                              rank_id=mp_rank,
-                              group=mp_group,
-                              use_calc_stream=True)
+                                  use_calc_stream=True))
+    if in_dygraph_mode():
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/meta_parallel/pp_utils/p2p_communication.py:433]: There is no task in static program?

File: python/paddle/distributed/fleet/meta_parallel/parallel_layers/pp_layers.py [hunk 4]
```diff
@@ -189,7 +189,7 @@ def forward(self, *args, **kwargs):
         # Users shouldn't call PipelineLayerChunk directly, since all logics relating with recompute
         # are in the forward function of PipelineLayer. Any directly call will bring unexpected
         # behavior under recompute circumstance.
-        raise NotImplementedError(
+        raise PermissionError(
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/meta_parallel/parallel_layers/pp_layers.py:192]: Thanks for your hard work!

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 5]
```diff
@@ -22,11 +22,47 @@
 from ..utils.hybrid_parallel_util import broadcast_sharding_parameters
 from ..utils.log_util import logger
 from ..meta_optimizers.dygraph_optimizer import HybridParallelOptimizer, HybridParallelGradScaler
-from .pp_utils import p2p_communication as p2p
+from paddle.fluid.framework import _in_legacy_dygraph, _non_static_mode, in_dygraph_mode
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]:
```
import paddle.fluid.framework as framework
framework.xxx
```

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 6]
```diff
@@ -22,11 +22,47 @@
...
+    return _REAL_PP_RANK == (_REAL_PP_WORLD_SIZE - 1)
+
+
+def set_virtual_pipeline_rank(rank):
+    global _VIRTUAL_PP_RANK
+    _VIRTUAL_PP_RANK = rank
+
+
+# p2p communication depends on the three functions above
+from .pp_utils import p2p_communication as p2p
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: Move upward?

**Review by gongweibao — CHANGES_REQUESTED** [review]
Some comments.

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 6 above)
Comment (FeixLiu) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: done

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 5 above)
Comment (FeixLiu) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: done

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 1 above)
Comment (FeixLiu) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: Has move all these checks into PipelineParallel

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 2 above)
Comment (FeixLiu) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: Ok, nice point

File: python/paddle/fluid/tests/unittests/collective/fleet/hybrid_parallel_pp_layer_with_virtual_stage.py [hunk 7]
```diff
@@ -87,7 +87,7 @@ def test_pipelayer_desc(self):
 
         try:
             model_chunks[0](paddle.to_tensor([1., 2.]))
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/collective/fleet/hybrid_parallel_pp_layer_with_virtual_stage.py:89]: raise not implemented error!

Comment (FeixLiu) [reply python/paddle/fluid/tests/unittests/collective/fleet/hybrid_parallel_pp_layer_with_virtual_stage.py:89]: done

**Review by gongweibao — APPROVED** [review]
LGTM

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 8]
```diff
@@ -364,3 +414,291 @@ def _optimizer_step(self):
...
+        self.scaler = scaler
+        self.data = data
+        self.total_loss = None
+        self.micro_batch_id = 0
+        self._forward_only = forward_only
+
+        # init some data buffers for interleave scheduler
+        self.input_tensors = [[] for _ in range(self.num_model_chunks)]
+        self.output_tensors = [[] for _ in range(self.num_model_chunks)]
+        self.output_tensor_grads = [[] for _ in range(self.num_model_chunks)]
```
Comment (sneaxiy) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:512]: Not quite important: `self.output_tensor_grads` is not necessary when `forward_only = True`.

**Review by sneaxiy — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
下个PR删除`RUN_SERIAL 1`参数
